x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/CacheBufferedIndexInput.java

```diff
@@ -214,7 +219,11 @@ private int readDirectly(long start, long end, byte[] buffer, int offset) throws
             while (remaining > 0) {
                 final int len = (remaining < copyBuffer.length) ? (int) remaining : copyBuffer.length;
                 int bytesRead = input.read(copyBuffer, 0, len);
-                System.arraycopy(copyBuffer, 0, buffer, offset + bytesCopied, len);
+                if (bytesRead == -1) {
+                    throw new EOFException(String.format(Locale.ROOT,
+                        "unexpected EOF reading [%d-%d] ([%d] bytes remaining) from %s", start, end, remaining, cacheFileReference));
+                }
+                System.arraycopy(copyBuffer, 0, buffer, offset + bytesCopied, bytesRead);
```

Comment:
Replaces `len` with `bytesRead` in case of a partial read. I think this was technically ok before since we would proceed to overwrite the junk bytes with correct ones, but it was confusing.